src/main/java/randoop/generation/OrienteeringSelection.java

```diff
@@ -0,0 +1,185 @@
+package randoop.generation;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.plumelib.util.CollectionsPlume;
+import randoop.sequence.ExecutableSequence;
+import randoop.sequence.Sequence;
+import randoop.util.Randomness;
+import randoop.util.SimpleList;
+
+/**
+ * Implements the Orienteering component, as described by the paper "GRT: Program-Analysis-Guided
+ * Random Testing" by Ma et. al (appears in ASE 2015):
+ * https://people.kth.se/~artho/papers/lei-ase2015.pdf .
+ *
+ * <p>Biases input selection towards sequences that have lower execution cost. Execution cost is
+ * measured by the number of method calls in a sequence and the time it takes to execute.
+ *
+ * <p>Our implementation of Orienteering differs from that described in the GRT paper in that we do
+ * not keep track of the execution time of an input sequence for each new run. Instead, we assume
+ * that an input sequence's execution time is equal to the execution time of its first run. We
+ * believe this assumption is reasonable since a sequence does not take any inputs (it is
+ * self-contained), so its execution time probably does not differ greatly between separate runs.
+ *
+ * <p>The GRT paper also does not describe how to handle input sequences that have an execution time
+ * of zero, such as one that only includes the assignment of a primitive type {@code byte byte0 =
+ * (byte)1;}. We assign these input sequences an execution time of 1 nanosecond to prevent division
+ * by zero when later computing weights.
+ */
+public class OrienteeringSelection extends InputSequenceSelector {
+  /** Map from a sequence to its weight. */
+  private final Map<Sequence, Double> weightMap = new HashMap<>();
+
+  /** Map from a sequence to the number of times it was selected. */
+  private final Map<Sequence, Integer> sequenceSelectionCount = new HashMap<>();
+
+  /**
+   * Cache for the square root of the number of method calls in a sequence of statements. Once
+   * computed for a given sequence, the value is never updated.
+   */
+  private final Map<Sequence, Double> sequenceMethodSizeSqrt = new HashMap<>();
+
+  /**
+   * Map from a sequence to its execution time in nanoseconds. Once computed for a given sequence,
+   * the value is never updated.
+   */
+  private final Map<Sequence, Long> sequenceExecutionTime = new HashMap<>();
+
+  /**
+   * Bias input selection towards lower-cost sequences. We first compute and update the weights of
+   * all the candidates within the candidate list before making our selection.
```

Comment:
"all the candidates" is incorrect: only those that haven't beer computed already are computed by this method.
However, I still don't see why this method needs to be called (or, maybe it only needs to be called once at the very beginning).  I would think that a candidate only gets added to the list when its weight is computed.  How can elements without a weight appear in the list?

Reply:
This is because Orienteering is selecting inputs for a method and a call to the input selection method will be affected by the type. Specifically, the candidate list that is passed in to

    public Sequence selectInputSequence(SimpleList<Sequence> candidates)

will contain different types for different calls. So Orienteering won't have an entire list of all possible inputs.
Thus, elements without a weight can appear in the list because it could be the case that the input selection method was called for a type that Orienteering hasn't looked at before. These sequences wouldn't be ones generated by Randoop during the test generation process but would be the initial seed sequences with no execution time.

Reply:
Please correct "all the candidates" to indicate it's only those that haven't been computed already.

Also remark that the only sequences whose weight has not been computed are initial seed sequences.

However, is it necessary to call this method frequently?  Would it be possible to, early in the run of Randoop, compute weights for all the initial seed sequences?  Then there would be an invariant that every sequence in the pool has a weight, and it would not be necessary to iterate over all the sequences looking for ones that need to have their weight updated.